Add 10 solutions for Rosalind tasks

dadlex opened the pull request.

dadlex committed:
[task1] add task count dna nucleotides

dadlex committed:
[task1] wrap input stream with try with resources

dadlex committed:
[task1] make task1 methods static

dadlex committed:
[task1] make input parameters final

dadlex committed:
[task1] make appropriate vars final

dadlex committed:
[Task 2] Add evaluation of hamming distance
[Task 3] Add evaluation of DNA reverse complement
[Task 4] Add evaluation of transcribed DNA

dadlex committed:
[Task 5] Add protein string creation out of rna
[Task 6] Add protein mass evaluation

dadlex committed:
[Task 7] add substring occurrences search
[Task 8] add gc content evaluation

dadlex committed:
[Task 9] add rna splicing

copy-paste of previously declared methods. these taks have nothing to do
with OOP, so I consider each task independent

dadlex committed:
[Task 10] add string permutation algorithm

dmitrii-adamian-at-epam commented on alexander_aphonin/countDnaNucleotides/Task1.java:
```diff
@@ -0,0 +1,40 @@
+    private static String readInputStreamIntoString(final InputStream inputStream)
+            throws IOException {
+        final StringBuilder resultStringBuilder = new StringBuilder();
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream))) {
```
Здесь можно использовать метод [Files.readAllLines()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#readAllLines(java.nio.file.Path))

dmitrii-adamian-at-epam commented on alexander_aphonin/countDnaNucleotides/Task1.java:
```diff
@@ -0,0 +1,40 @@
+                    .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+            System.out.println(l.toString());
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Здесь лучше вообще не ловить исключения - результат точно такой же, а кода меньше.

dadlex replied:
Не до конца понял имеется в виду только FileNotFoundException или IOException тоже (с переносом в объявление функции). Поскольку обработки здесь все равно никакой нет, выбрал первый вариант.

dmitrii-adamian-at-epam replied:
Вообще в таких программках проще ставить throws Exception и не заморачиваться вообще.
Но оставьте как есть.

dmitrii-adamian-at-epam commented on alexander_aphonin/dnaReverseComplement/Task3.java:
```diff
@@ -0,0 +1,29 @@
+public class Task3 {
+    public static Map<String, String> nucleotideComplements;
+    static {
+        nucleotideComplements = new HashMap<>();
```
Блоки инициализации обычно не используют без явной необходимости, без них читаемость лучше.
Здесь можно сделать так:
```java
public static final Map<String, String> nucleotideComplements = Map.of(
    "A", "T",
    "C", "G",
    "T", "A",
    "G", "C",
);
```

dmitrii-adamian-at-epam commented on alexander_aphonin/dnaReverseComplement/Task3.java:
```diff
@@ -0,0 +1,29 @@
+        // we can ultimately skip creation of just reversed string and make a stream out of reversed string here, but
+        // it hurts readability and as we only have 1000bp at max doesn't cost much.
+        final String reversedDnaString = new StringBuilder(dnaString).reverse().toString();
+        return Stream.of(reversedDnaString.split(""))
```
Для больших строк эта операция создаст очень много объектов, соответственно увеличит потребление памяти. Лучше использовать операции с символами. Каждый символ использует 2 байта, а даже пустая строка занимает 40 байт.
Можете не исправлять для этой задачи, но учитывайте это в будущем. Это один из главных факторов избыточного потребления памяти при работе с Java.

dmitrii-adamian-at-epam commented on alexander_aphonin/hammingDistance/Task2.java:
```diff
@@ -0,0 +1,20 @@
+public class Task2 {
+
+    public static int evalHammingDistance(final String dnaSeq1, final String dnaSeq2)
+            throws IllegalArgumentException {
```
IllegalArgumentException можно не объявлять в сигнатуре метода - это не проверяемое исключение.

dmitrii-adamian-at-epam commented on alexander_aphonin/printPermutations/Task10.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public int getPermutationsAmount() {
```
Геттер и сеттер нигде не используются. Может быть, их лучше удалить?

dadlex replied:
Они используются в 33 строке, так как мы обращаемся и изменяем счетчик. можно и без методов, конечно, но если добавится логика, что при изменении или обращении нужные побочные эффекты, то это может привести к ошибкам

dmitrii-adamian-at-epam commented on alexander_aphonin/proteinMassCalculation/Task6.java:
```diff
@@ -0,0 +1,41 @@
+        // in prod we should check that it's AUCG string and that the length is divisible by 3.
+        // we can also use stop sequences here, but it's not asked.
+        return Stream.of(proteinString.split(""))
+                .map(protein -> monoisotopicMasses.get(protein))
```
Здесь можно сразу использовать`mapToDouble()`.

dmitrii-adamian-at-epam commented on alexander_aphonin/proteinMassCalculation/Task6.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public static Double calculateProteinMass(final String proteinString) {
+        // in prod we should check that it's AUCG string and that the length is divisible by 3.
```
Комментарий кажется относящимся не к этому коду.

dmitrii-adamian-at-epam commented on alexander_aphonin/proteinStringFromRna/Task5.java:
```diff
@@ -0,0 +1,86 @@
+        // in prod we should check that it's AUCG string and that the length is divisible by 3.
+        // we can also use stop sequences here, but it's not asked.
+        return IntStream
+                .iterate(0, i -> i + 3)
```
Для числа 3 лучше использовать именованную константу `CODON_LENGTH`.

dadlex replied:
Однозначно

dmitrii-adamian-at-epam commented on alexander_aphonin/transcribeDna/Task4.java:
```diff
@@ -0,0 +1,15 @@
+
+public class Task4 {
+    public static String getTranscribedRna(final String dnaString) {
+        return Stream.of(dnaString.split(""))
```
Можно использовать [String.replace()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#replace(char,char))

dadlex replied:
Точно

dmitrii-adamian-at-epam commented on alexander_aphonin/substringOccurrences/Task7.java:
```diff
@@ -0,0 +1,17 @@
+        Pattern substringPattern = Pattern.compile("(?=" + substring + ")");
+        Matcher substringMatcher = substringPattern.matcher(dnaString);
+        // for some reason the next piece doesn't work. Would appreciate the comment.
+        // .map(occurrence -> occurrence.start() + 1)
```
К сожалению, не могу прокомментировать почему не работает.

dadlex replied:
Результатом становится только последнее совпадение, хотя внутрь map мы попадаем три раза. Почему toList три числа не смог в список сложить - неясно

dmitrii-adamian-at-epam commented on alexander_aphonin/rnaSplicing/Task9.java:
```diff
@@ -0,0 +1,143 @@
+        proteinAbbreviations.put("GGA", "G");
+        proteinAbbreviations.put("GGG", "G");
+    }
+    public static Map<String, String> readFastaFile(final String fastaFilename) {
```
Можно этот код переиспользовать в двух задачах?

dadlex replied:
По-хорошему весь этот код нужно реструктурировать, чтобы тут ноги не переломать, но хорошо, Utils класс можно завести

dadlex replied:
Я бы сказал скорее нет, потому что восьмая задача подразумевает считывание однородных данных, а девятая данных с заголовком по сути

dmitrii-adamian-at-epam commented on alexander_aphonin/rnaSplicing/Task9.java:
```diff
@@ -0,0 +1,143 @@
+
+        // we exclude the dna string itself. to only erase introns.
+        fileData.entrySet().stream()
+                .forEach(intron -> dnaSequenceEntry.setValue(dnaSequenceEntry.getValue().replaceAll(intron.getValue(),"")));
```
В этой строке слишком много операций, она плохо читается.
Также можно использовать [Map.forEach()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#forEach(java.util.function.BiConsumer))

dmitrii-adamian-at-epam commented on alexander_aphonin/rnaSplicing/Task9.java:
```diff
@@ -0,0 +1,143 @@
+    }
+    public static Map<String, String> readFastaFile(final String fastaFilename) {
+        //unlike earlier task we do need the ordering guarantees, hence the linkedHashMap
+        Map<String, String> dnaStrings = new LinkedHashMap<>(); // <id, data>
```
Использовать LinkedHashMap для сохранения порядка - решение хоть и допустимое, но плохо читаемое.
Обычно лучше сделать две связанные структуры данных - List и Map (неупорядоченный), можно сделать их полями одного класса. Чуть больше кода, зато "что хотел сказать автор" сразу очевидно.

dadlex replied:
не понял схему про List и Map, если можно киньтесь ссылкой. А так по сути здесь достаточно обычной хэш мапы + один Pair<String,String> на header id, header data (первоначальная dna строка) и все это в обертке, но не знаю есть ли смысл.
А на тему неочевидности, что хотел сказать автор - ну да, неочевидно, согласен. особенно учитывая, что id для задачи вообще не нужны. Поэтому переделал на обычный лист. Если нужно все-таки сделать комбинацию, то расскажите чуть подробнее, пожалуйста

dmitrii-adamian-at-epam replied:
Примерно так:
```java
OperationResult processSomething() {
   ...
}

private static final class OperationResult {
    List<KeyClass> resultObjects;
    Map<KeyClass, ValueClass> objectsByKey;
}
```

Тут сразу понятно где упорядочено, а где мапа с доступом по ключу. Но несколько многословно.
Можете не делать, не обязательно.

dadlex replied:
Понял, да, действительно показательно, спасибо

dmitrii-adamian-at-epam reviewed (commented):
У меня несколько стилистических замечаний, в остальном все хорошо.
После исправления запрашивайте ревью еще раз.

dadlex committed:
[Task 1] change BufferedReader to Files.readAllLines()
[Task 2] remove IllegalArgumentException  from signature
[Task 3] remove static init block
[Task 4] use String.replace() instead of streams
[Task 5] change random int to a const
[Task 6] remove redundant streams api call
[Task 9] change map to list
    change forEach to reduce to work around immutable string

dadlex committed:
[Task 9] change LinkedList to ArrayList

dmitrii-adamian-at-epam reviewed (approved):
Зачтено, выбирайте проект.
